server/subscription/manager.go

```diff
@@ -4,54 +4,51 @@ import "sync"
 
 type Manager struct {
 	subscriptions map[string][]Subscriber
-	mutex         sync.RWMutex
+	mutex         sync.Mutex
 }
 
 func NewManager() *Manager {
 	return &Manager{
 		subscriptions: make(map[string][]Subscriber),
-		mutex:         sync.RWMutex{},
+		mutex:         sync.Mutex{},
 	}
 }
 
-func (m *Manager) Subscribe(resourceID string, subscriber Subscriber) {
+func (m *Manager) getSubscribers(resourceID string) []Subscriber {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-
-	array := make([]Subscriber, 0)
-
-	if existingArray, ok := m.subscriptions[resourceID]; ok {
-		array = existingArray
-	}
-
-	array = append(array, subscriber)
-	m.subscriptions[resourceID] = array
+	return m.subscriptions[resourceID]
 }
 
-func (m *Manager) Unsubscribe(resourceID string, subscriptionID string) {
+func (m *Manager) setSubscribers(resourceID string, subscribers []Subscriber) {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
+	m.subscriptions[resourceID] = subscribers
+}
 
-	array, exists := m.subscriptions[resourceID]
-	if !exists {
-		return
-	}
+func (m *Manager) Subscribe(resourceID string, subscriber Subscriber) {
+	subscribers := append(m.getSubscribers(resourceID), subscriber)
+	m.setSubscribers(resourceID, subscribers)
```

Comment:
Don't you need to lock the mutex here to prevent read-write race conditions here? Go will panic in case of read-write concurrency: https://stackoverflow.com/questions/36167200/how-safe-are-golang-maps-for-concurrent-read-write-operations

Same for PublishUpdate

Reply:
Originally each function that needed to access the map would manage the lock itself. That code was causing issues for some reason, so I moved the usage of the lock to the [getSubscribers](https://github.com/kubeshop/tracetest/pull/3098/files/52bc8cfbad5e4cf9d39692d3cf5c4f9744b75957#diff-e5da02bda163f4561d8b7d1d92e295305327f3afd4235a8de026fd2a5f31da7aR17-R19) function and the [setSubscribers](https://github.com/kubeshop/tracetest/pull/3098/files/52bc8cfbad5e4cf9d39692d3cf5c4f9744b75957#diff-e5da02bda163f4561d8b7d1d92e295305327f3afd4235a8de026fd2a5f31da7aR23-R25) function, so effectively every read and write is lock protected